File: packages/health_connector_hk_ios/ios/health_connector_hk_ios/Sources/health_connector_hk_ios/handlers/DiastolicBloodPressureHandler.swift
import Foundation
import HealthKit

/// Handler for diastolic blood pressure (instant quantity type)
///
/// **Data Type:** Diastolic Blood Pressure
/// **Record Pattern:** Instant (single timestamp)
/// **Aggregation:** Average, Min, Max
/// **HealthKit Type:** HKQuantityTypeIdentifier.bloodPressureDiastolic
///
/// **Note:** When writing a DiastolicBloodPressureRecord on iOS, it creates
/// a standalone quantity sample, not a correlation. For complete blood
/// pressure readings (systolic + diastolic together), use BloodPressureRecord.

struct DiastolicBloodPressureHandler: HealthKitQuantityHandler {
    // MARK: - HealthKitTypeHandler

    static var supportedType: HealthDataTypeDto {
        return .diastolicBloodPressure
    }

    static var category: HealthKitDataCategory {
        return .quantitySample
    }

    // MARK: - HealthKitSampleHandler

    static func toDTO(_ sample: HKSample) throws -> HealthRecordDto {
        guard let quantitySample = sample as? HKQuantitySample else {
            throw HealthConnectorErrors.invalidArgument(
                message: "Expected HKQuantitySample, got \(type(of: sample))"
            )
        }
        guard let dto = quantitySample.toDiastolicBloodPressureRecordDto() else {
            throw HealthConnectorErrors.invalidArgument(message: "Failed to convert HKQuantitySample to DiastolicBloodPressureRecordDto")
        }
        return dto
    }

    static func toHealthKit(_ dto: HealthRecordDto) throws -> HKSample {
        guard let diastolicDto = dto as? DiastolicBloodPressureRecordDto else {
            throw HealthConnectorErrors.invalidArgument(
                message: "Expected DiastolicBloodPressureRecordDto, got \(type(of: dto))"
            )
        }
        return try diastolicDto.toHealthKit()
    }

    static func getSampleType() -> HKSampleType {
        return HKQuantityType.quantityType(forIdentifier: .bloodPressureDiastolic)!
    }

    static func extractTimestamp(_ dto: HealthRecordDto) throws -> Int64 {
        guard let diastolicDto = dto as? DiastolicBloodPressureRecordDto else {
            throw HealthConnectorErrors.invalidArgument(
                message: "Expected DiastolicBloodPressureRecordDto, got \(type(of: dto))"
            )
        }
        return diastolicDto.time
    }

    // MARK: - HealthKitQuantityHandler (Aggregation Support)

    static func toStatisticsOptions(_ metric: AggregationMetricDto) throws -> HKStatisticsOptions {
        switch metric {
        case .avg:
            return .discreteAverage
        case .min:
            return .discreteMin
        case .max:
            return .discreteMax
        case .sum, .count:
            throw HealthConnectorErrors.invalidArgument(
                message: "Aggregation metric '\(metric)' not supported for diastolic blood pressure (discrete data)",
                details: "Supported metrics: avg, min, max"
            )
        }
    }

    static func extractAggregateValue(
        from statistics: HKStatistics,
        metric: AggregationMetricDto
    ) throws -> MeasurementUnitDto {
        let quantity: HKQuantity? = switch metric {
        case .avg:
            statistics.averageQuantity()
        case .min:
            statistics.minimumQuantity()
        case .max:
            statistics.maximumQuantity()
        case .sum, .count:
            throw HealthConnectorErrors.invalidArgument(
                message: "Aggregation metric '\(metric)' not supported for diastolic blood pressure",
                details: "Supported metrics: avg, min, max"
            )
        }

        guard let quantity else {
            throw HealthConnectorErrors.invalidArgument(
                message: "No aggregation result for metric '\(metric)'",
                details: "Statistics returned nil for bloodPressureDiastolic"
            )
        }

        return quantity.toPressureDto()
    }
}

File: packages/health_connector_hc_android/android/src/main/kotlin/com/phamtunglam/health_connector_hc_android/handlers/DiastolicBloodPressureHandler.kt
package com.phamtunglam.health_connector_hc_android.handlers

import androidx.health.connect.client.aggregate.AggregateMetric
import androidx.health.connect.client.records.BloodPressureRecord
import androidx.health.connect.client.records.Record
import androidx.health.connect.client.units.Pressure
import com.phamtunglam.health_connector_hc_android.mappers.toDiastolicDto
import com.phamtunglam.health_connector_hc_android.mappers.toDto
import com.phamtunglam.health_connector_hc_android.mappers.toHealthConnect
import com.phamtunglam.health_connector_hc_android.pigeon.AggregationMetricDto
import com.phamtunglam.health_connector_hc_android.pigeon.DiastolicBloodPressureRecordDto
import com.phamtunglam.health_connector_hc_android.pigeon.HealthDataTypeDto
import com.phamtunglam.health_connector_hc_android.pigeon.HealthRecordDto
import com.phamtunglam.health_connector_hc_android.pigeon.MeasurementUnitDto
import com.phamtunglam.health_connector_hc_android.pigeon.PressureDto
import com.phamtunglam.health_connector_hc_android.pigeon.PressureUnitDto
import kotlin.reflect.KClass

/**
 * Handler for Diastolic Blood Pressure health data type.
 *
 * Characteristics:
 * - Category: Instant record (single timestamp)
 * - Aggregation: Supports AVG, MIN, MAX (using BloodPressureRecord.DIASTOLIC_* metrics)
 * - Health Connect Type: BloodPressureRecord (with systolic set to 0)
 * - Note: Writes create a BloodPressureRecord with systolic=0
 */
object DiastolicBloodPressureHandler : InstantRecordHandler, AggregationSupportingHandler {
    override val supportedType: HealthDataTypeDto = HealthDataTypeDto.DIASTOLIC_BLOOD_PRESSURE

    override fun toDto(record: Record): HealthRecordDto {
        require(record is BloodPressureRecord) {
            "Expected BloodPressureRecord, got ${record::class.simpleName}"
        }
        return record.toDiastolicDto()
    }

    override fun toHealthConnect(dto: HealthRecordDto): Record {
        require(dto is DiastolicBloodPressureRecordDto) {
            "Expected DiastolicBloodPressureRecordDto, got ${dto::class.simpleName}"
        }
        return dto.toHealthConnect()
    }

    override fun getRecordClass(): KClass<out Record> = BloodPressureRecord::class

    override fun supportedAggregations(): List<AggregationMetricDto> {
        return listOf(
            AggregationMetricDto.AVG,
            AggregationMetricDto.MIN,
            AggregationMetricDto.MAX
        )
    }

    override fun toAggregateMetric(metric: AggregationMetricDto): AggregateMetric<*> {
        return when (metric) {
            AggregationMetricDto.AVG -> BloodPressureRecord.DIASTOLIC_AVG
            AggregationMetricDto.MIN -> BloodPressureRecord.DIASTOLIC_MIN
            AggregationMetricDto.MAX -> BloodPressureRecord.DIASTOLIC_MAX
            else -> error("Unsupported aggregation metric $metric for DiastolicBloodPressure. Supported: AVG, MIN, MAX")
        }
    }

    override fun extractAggregateValue(
        aggregatedValue: Any?,
        metric: AggregationMetricDto
    ): MeasurementUnitDto {
        val pressure = aggregatedValue as? Pressure
        return pressure?.toDto() ?: PressureDto(value = 0.0, unit = PressureUnitDto.MILLIMETERS_OF_MERCURY)
    }
}
